src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1438,15 +1437,7 @@ uct_dc_mlx5_dci_keepalive_handle_failure(uct_dc_mlx5_iface_t *iface,
     }
 
     ep = ucs_derived_of(op->ep, uct_dc_mlx5_ep_t);
-
-    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        ucs_assert(ep != iface->tx.fc_ep);
-        uct_dc_mlx5_iface_set_ep_failed(iface, ep, cqe, txwq, ep_status);
-    } else {
-        /* ep has another dci assigned to post operations, which should be
-         * restarted too */
-        uct_dc_mlx5_ep_handle_failure(ep, cqe, ep_status);
-    }
+    uct_dc_mlx5_ep_handle_failure(ep, dci_index, cqe, ep_status);
```

Comment:
maybe just never call uct_dc_mlx5_ep_handle_failure - always call uct_dc_mlx5_iface_set_ep_failed?

Reply:
why? i would expect it would be returned when we have completion (success) on the normal dci

Reply:
yeah, right
it should be released on succes/fail completions

Reply:
done